borsh-derive-internal/src/attribute_helpers.rs

```diff
@@ -18,3 +23,120 @@ pub fn contains_initialize_with(attrs: &[Attribute]) -> Option<Path> {
 
     None
 }
+pub fn check_use_discriminant(derive_input: DeriveInput) -> Result<Option<bool>, TokenStream> {
```

Comment:
Correction:
this should be split in 2 methods:
```rust
pub fn check_item_attributes(derive_input: &DeriveInput) -> Result<(), proc_macro2::TokenStream> {
...
pub fn contains_use_discriminant(attrs: &[Attribute]) -> Result<Option<bool>, proc_macro2::TokenStream>
```
the `proc_macro2::TokenStream` can then be cast to `proc_macro::TokenStream` in `borsh-derive/src/lib.rs`:
```rust
   match check_item_attributes(&derive_input) {
        Ok(..) => {},
        Err(value) => return value.into(),
    };

```

Reply:
According to this comment and another one where you say to move this check in this function I got function that check if we call our macros on struct or enum, but I need to do this check only if have `use_discriminant=<bool>` attribute. Also, I am not sure in which context  `check_item_attributes` should be called since I can't call it inside `contains_use_discriminant` because it lacks of context and in some another context where DeriveInput is available it checks  if we call our macros on struct or enum which is not I want. 
```
pub fn check_item_attributes(derive_input: &DeriveInput) -> Result<(), proc_macro2::TokenStream> {
    if let syn::Data::Struct(ref _data) = derive_input.data {
        return Err(TokenStream::from(
            syn::Error::new(
                derive_input.ident.span(),
                "borsh (use_discriminant=<bool>) does not support structs",
            )
            .to_compile_error(),
        ));
    }
    Ok(())
}
```

Reply:
```rust
// validation
pub fn check_item_attributes(derive_input: &DeriveInput) -> Result<(), TokenStream> {
    for attr in &derive_input.attrs {
        // attr.path can technically be only `borsh`, `borsh_init` and `borsh_skip`
        // which is checked by compiler
        // using `borsh_skip` here is error

        if attr.path().is_ident("borsh_skip") {
            return Err(TokenStream::from(
                syn::Error::new(
                    derive_input.ident.span(),
                    "`borsh_skip` is not allowed as derive input attribute",
                )
                .to_compile_error(),
            ));
        }
        if attr.path().is_ident("borsh") {
            attr.parse_nested_meta(|meta| {
                // this line changes to
                // if !meta.path.is_ident(USE_DISCRIMINANT) && !meta.path.is_ident(INIT) {
                // later, when `borsh_init` becomes `borsh(init)`
                if !meta.path.is_ident(USE_DISCRIMINANT) {
                    return Err(syn::Error::new(
                        derive_input.ident.span(),
                        "`use_discriminant` is the only supported attribute for `borsh`",
                    ));
                }
                if meta.path.is_ident(USE_DISCRIMINANT) {
                    let _expr: Expr = meta.value()?.parse()?;
                    if let syn::Data::Struct(ref _data) = derive_input.data {
                        return Err(syn::Error::new(
                            derive_input.ident.span(),
                            "borsh(use_discriminant=<bool>) does not support structs",
                        ));
                    }
                }

                Ok(())
            })
            .map_err(|err| TokenStream::from(err.to_compile_error()))?;
        }
    }
    Ok(())
}
```
```rust
// gets what you need
pub fn contains_use_discriminant(attrs: &[Attribute]) -> Result<Option<bool>, TokenStream> {
    let mut result = None;
    for attr in attrs {
        if attr.path().is_ident("borsh") {
            attr.parse_nested_meta(|meta| {
                if meta.path.is_ident(USE_DISCRIMINANT) {
                    let value_expr: Expr = meta.value()?.parse()?;
                    let value = value_expr.to_token_stream().to_string();
                    // this goes to contains_use_discriminant
                    match value.as_str() {
                        "true" => {result = Some(true);},
                        "false" => {result = Some(false)},
                        _ => {
                            return Err(syn::Error::new(
                                value_expr.span(),
                                "`use_discriminant` accept only `true` or `false`",
                            ));
                        }
                    };
                }

                Ok(())
            })
            .map_err(|err| TokenStream::from(err.to_compile_error()))?;
        }
    }
    Ok(result)
}
```
the actual check `"use_discriminant is the only supported attribute for borsh",` from this [commit](https://github.com/near/borsh-rs/pull/148/commits/e19f9b98d7c65e9b597249f610e54fd57a143a5d) doesn't work
as it only checks some of `Meta` variants